### PR TITLE
[IMP] account,account_edi_ubl_cii: Send & Print improvements

### PR DESCRIPTION
**Pull request by clbr-odoo**
1. [TECHNICAL] Move send & print files into wizard folder when it's not the case

2. [TECHNICAL] More testing: test cron + cron in multicompany setup

3. Make 'invoices.zip' translatable

4. Use the bus to notify when cron is done (and maybe when it's started ? )

task-id: 3568549

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -4181,8 +4199,28 @@ def _cron_account_move_send(self, job_count=10):
...
+            moves_map = {
+                partner: moves.filtered(lambda m: m.send_and_print_values['sp_partner_id'] == partner.id)
+                for partner in sp_partners
+            }
+
             self.env['account.move.send']._process_send_and_print(moves)
 
+            notifications = []
+            for partner, partner_moves in moves_map.items():
+                partner_moves_error = partner_moves.filtered(lambda m: m.send_and_print_values and m.send_and_print_values.get('error'))
```
Comment (smetl) [inline addons/account/models/account_move.py:4214]: `send_and_print_values` is always set here
```suggestion
                partner_moves_error = partner_moves.filtered(lambda m: m.send_and_print_values.get('error'))
```

File: addons/account/static/src/services/account_notification_service.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        bus_service.subscribe("account_notification", ({ message, sticky, title, type, action_button}) => {
+            const buttons = [{
+                name: action_button.name,
+                primary: false,
+                onClick: () => {
+                    action.doAction({
+                        name: _t(action_button.action_name),
+                        type: 'ir.actions.act_window',
+                        res_model: action_button.model,
+                        domain: [["id", "in", (action_button.res_ids)]],
```
Comment (smetl) [inline addons/account/static/src/services/account_notification_service.js:20]:
```suggestion
                        domain: [["id", "in", action_button.res_ids]],
```

File: addons/account/static/src/services/account_notification_service.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                        views: [[false, 'list'],[false, 'form']],
```
Comment (smetl) [inline addons/account/static/src/services/account_notification_service.js:21]:
```suggestion
                        views: [[false, 'list'], [false, 'form']],
```

File: addons/account/static/src/services/account_notification_service.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+                        target: 'current',
+                    });
+                },
+            }];
+            notification.add(message, { sticky, title, type, buttons });
+        });
+        bus_service.start();
```
Comment (smetl) [inline addons/account/static/src/services/account_notification_service.js:28]: Are you sure you need to start this service yourself? :thinking:

File: addons/account/tests/test_account_move_send.py [hunk 5]
```diff
@@ -934,3 +934,53 @@ def test_with_draft_invoices(self):
...
+        wizard_2 = self.create_send_and_print(invoice_2_1 + invoice_2_2)
+        wizard_2.checkbox_download = False
+        wizard_2.action_send_and_print()
+
+        invoices = invoice_1_1 + invoice_1_2 + invoice_2_1 + invoice_2_2
+        self.assertFalse(invoices.invoice_pdf_report_id)
+        self.assertEqual(invoices.mapped(lambda inv: bool(inv.send_and_print_values)), [True] * len(invoices))
+        self.env.ref('account.ir_cron_account_move_send').method_direct_trigger()  # force processing
+        self.assertEqual(len(invoices.invoice_pdf_report_id), 4)
+        self.assertEqual(invoices.mapped(lambda inv: inv.send_and_print_values), [False] * len(invoices))
```
Comment (smetl) [inline addons/account/tests/test_account_move_send.py:957]:
```suggestion
        self.assertTrue(all(inv.invoice_pdf_report_id for inv in invoices))
        self.assertTrue(all(not inv.send_and_print_values for inv in invoices))
```

File: addons/account/tests/test_account_move_send.py [hunk 6]
```diff
@@ -934,3 +934,53 @@ def test_with_draft_invoices(self):
...
+        wizard = self.create_send_and_print(invoice_success + invoice_error)
+        wizard.checkbox_download = False
+        sp_partner = self.env.user.partner_id
+        wizard.action_send_and_print()
+
+        def _hook_invoice_document_before_pdf_report_render(self, invoice, invoice_data):
+            if invoice == invoice_error:
+                invoice_data['error'] = 'blblblbl'
+
+        self.assertEqual(invoice_success.send_and_print_values.get('sp_partner_id'), sp_partner.id)
```
Comment (smetl) [inline addons/account/tests/test_account_move_send.py:971]:
```suggestion
        self.assertTrue(invoice_success.send_and_print_values)
        self.assertEqual(invoice_success.send_and_print_values.get('sp_partner_id'), sp_partner.id)
```

File: addons/account/tests/test_account_move_send.py [hunk 7]
```diff
@@ -934,3 +934,53 @@ def test_with_draft_invoices(self):
...
+
+        with patch(
+                'odoo.addons.account.wizard.account_move_send.AccountMoveSend._hook_invoice_document_before_pdf_report_render',
+                _hook_invoice_document_before_pdf_report_render,
+        ):
+            self.env.ref('account.ir_cron_account_move_send').method_direct_trigger()  # force processing
+
+        bus = self.env['bus.bus'].sudo().search(
+            [('channel', 'like', f"\"res.partner\",{sp_partner.id}")],
```
Comment (smetl) [inline addons/account/tests/test_account_move_send.py:980]:
```suggestion
            [('channel', 'like', f'"res.partner",{sp_partner.id}')],
```

File: addons/account/wizard/account_move_send.py [hunk 8]
```diff
@@ -672,8 +672,13 @@ def _process_send_and_print(self, moves, wizard=None, allow_fallback_pdf=False,
...
 
         # Update send and print values of moves
-        for move in moves:
-            if move.send_and_print_values:
+        for move, move_data in moves_data.items():
+            if from_cron and move_data.get('error'):
+                move.send_and_print_values = {
+                    **move.send_and_print_values,
+                    'error': True,
+                }
```
[the remark this answers is not in the document]
Comment (clbr-odoo) [reply addons/account/wizard/account_move_send.py:680]: Indeed.

File: addons/account/models/account_move.py (hunk 1 above)
Comment (clbr-odoo) [reply addons/account/models/account_move.py:4214]: Yes for the one above, but no for this one: the `_process_send_and_print` changes the `send_and_print_values` if everything is successful to False.

https://github.com/odoo/odoo/pull/149717/files#diff-69133af9cc727cff6007b627417e589681d05648c0902449a6e28506d0f422aeR682

File: addons/account/static/src/services/account_notification_service.js (hunk 4 above)
Comment (clbr-odoo) [reply addons/account/static/src/services/account_notification_service.js:28]: Looks like no :thinking: 
I used another implementation of such notification as example that had it, but some others don't... most likely it just need to be started once for the instance.

File: addons/account/models/account_move.py [hunk 9]
```diff
@@ -4158,6 +4158,24 @@ def _cron_account_move_send(self, job_count=10):
...
+                {
+                    'type': 'success' if is_success else 'warning',
+                    'title': _('Invoices sent') if is_success else _('Invoices in error'),
+                    'message': _('Invoices sent successfully.') if is_success else _(
+                        "One or more invoices couldn't be processed."),
+                    'action_button': {
+                        'name': _('Open'),
+                        'action_name': _('Sent invoices') if is_success else _('Invoices in error'),
+                        'model': 'account.move',
+                        'res_ids': moves.ids,
```
[the remark this answers is not in the document]
Comment (clbr-odoo) [reply addons/account/models/account_move.py]: I added that part in a separate commit so you see what it looks like. Imo it's even worse and I'd prefer to drop it :rofl:

**Comment by clbr-odoo** [comment]
@smetl Thanks for the review, interesting as always ! 
There are two left open conversations, the rest is fixed :slightly_smiling_face:

**Review by smetl — APPROVED** [review]

File: addons/account/models/account_move.py [hunk 10]
```diff
@@ -4181,8 +4199,28 @@ def _cron_account_move_send(self, job_count=10):
...
+            for partner, partner_moves in moves_map.items():
+                partner_moves_error = partner_moves.filtered(lambda m: m.send_and_print_values and m.send_and_print_values.get('error'))
+                if partner_moves_error:
+                    notifications.append(get_account_notification(partner, partner_moves_error, False))
+                partner_moves_success = partner_moves - partner_moves_error
+                if partner_moves_success:
+                    notifications.append(get_account_notification(partner, partner_moves_success, True))
+                partner_moves_error.send_and_print_values = False
+
+            self.env['bus.bus']._sendmany(notifications)
```
Comment (smetl) [inline addons/account/models/account_move.py:4222]: @odoo/rd-security This is executed from a CRON job. What is executed could come from multiple users. That's why we need to send many notifications to multiple users.

**Comment by smetl** [comment]
@robodoo rebase-ff

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**